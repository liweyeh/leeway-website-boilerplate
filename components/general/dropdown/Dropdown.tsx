import React, { useState, useEffect } from 'react';
import { NavLink, Button } from '@components';
import { LinkType } from '@types';
import styles from './dropdown.module.scss';

interface DropdownProps {
	children: React.ReactNode;
	menuOptions: LinkType[];
	customStyles?: string;
}

export const Dropdown = ({ children, menuOptions, customStyles }: DropdownProps) => {
	const [isOpen, setIsOpen] = useState(false);

	useEffect(() => {
		const closeSelect = (e: MouseEvent) => {
			if (e.defaultPrevented) return;
			setIsOpen(false);
		};
		document.addEventListener('click', closeSelect);

		return () => {
			document.removeEventListener('click', closeSelect);
		};
	}, []);

	return (
		<div data-cy="dropdown" className={[styles['dropdown'], customStyles].join(' ')}>
			<Button
				customStyles={styles['dropdown-button']}
				aria-label="dropdown menu"
				onClick={e => {
					e.stopPropagation();
					setIsOpen(prev => !prev);
				}}
			>
				{children}
			</Button>
			<div
				onClick={e => {
					e.preventDefault();
				}}
				className={[styles['dropdown-menu'], !isOpen && styles['dropdown-menu-hidden']].join(' ')}
			>
				{menuOptions.map((option, ind) => {
					const { displayName, href, isExternal } = option;
					return (
						<NavLink
							key={`${displayName} - ${ind}`}
							href={href}
							isExternal={isExternal}
							onClick={() => {
								setIsOpen(false);
							}}
						>
							<p>{displayName}</p>
						</NavLink>
					);
				})}
			</div>
		</div>
	);
};
