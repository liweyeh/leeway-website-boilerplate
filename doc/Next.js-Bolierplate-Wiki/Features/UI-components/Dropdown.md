[[_TOC_]]

# Overview

- Purpose: A dropdown navigation menu
- Type: General UI
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/dropdown)
- Created by: Liwei Yeh <liwei.yeh@adasasistemas.com>
- Created time: 23/08/2022

# Props

\* = required

- children<`React.ReactNode`>\*: Content to be displayed in the dropdown menu
- menuOptions<`boolean`>: Options for the dropdown
- customStyles<`string`>: Custom styles, expecting scss modules

# Usage

```tsx
import styles from './mainLayout.module.scss';

export const MainLayout = () => {
	const links = [
		{ href: '/', isExternal: false, displayName: 'Home' },
		{ href: '/theme', isExternal: false, displayName: 'Theme' },
	];

	return (
		<div className={styles.container}>
			<div>
				<Dropdown menuOptions={links} customStyles={styles.menu}>
					<div className={styles['menu-icon']}></div>
				</Dropdown>
			</div>
		</div>
	);
};
```
