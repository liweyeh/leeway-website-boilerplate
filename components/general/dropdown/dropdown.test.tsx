import { render } from '@testing-library/react';
import { Dropdown } from './Dropdown';
import '@testing-library/jest-dom';

const setup = () => {
	const { asFragment } = render(
		<Dropdown
			menuOptions={[
				{
					href: '/',
					isExternal: false,
					displayName: 'home',
				},
			]}
		>
			hi
		</Dropdown>,
	);
	return { asFragment };
};

test('test snapshot', () => {
	const { asFragment } = setup();
	expect(asFragment()).toMatchSnapshot();
});
