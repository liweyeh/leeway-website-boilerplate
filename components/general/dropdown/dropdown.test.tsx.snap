// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown "
    data-cy="dropdown"
  >
    <button
      aria-label="button"
      class="button-default  dropdown-button"
      type="button"
    >
      hi
    </button>
    <div
      class="dropdown-menu dropdown-menu-hidden"
    >
      <a
        aria-current="false"
        class="link"
        href="/"
      >
        <p>
          home
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
